Rubberduck.Parsing/VBA/VBACodeStringParser.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Atn;
+using Antlr4.Runtime.Tree;
+using NLog;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Symbols.ParsingExceptions;
+
+namespace Rubberduck.Parsing.VBA
+{
+    /// <summary>
+    /// The class is meant to be used for where we need to do rewriting without actually
+    /// rewriting an existing module in a VBA project. The result is completely free-floating
+    /// and it is the caller's responsibility to persist the finalized results into one of
+    /// actual code. This is useful for creating a code preview for refactoring.
+    /// </summary>
+    public class VBACodeStringParser
+    {
+        public enum ParserMode
+        {
+            Default,
+            Sll,
+            Ll
+        }
+
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+        private readonly CommonTokenStream _tokenStream;
+        private readonly VBAParser _parser;
+        private readonly string _moduleName;
+        private readonly ParserMode _mode;
+
+        /// <summary>
+        /// Parse a given string representing a module code. The code passed in must be
+        /// a valid module body, containing valid declarations or complete procedures. 
+        /// Code snippets are not valid. 
+        /// </summary>
+        /// <param name="moduleName">For logging purpose, provide descritpive name for where the virtual module is used</param>
+        /// <param name="moduleCodeString">A string containing valid module body.</param>
+        /// <param name="mode">Indicates what parser mode to use. By default we use SLL fallbacking to LL. When mode is explicitly specified, only that mode will be used with no fallback.</param>
+        public VBACodeStringParser(string moduleName, string moduleCodeString, ParserMode mode = ParserMode.Default)
+        {
+            var tokenStreamProvider = new SimpleVBAModuleTokenStreamProvider();
+            _tokenStream = tokenStreamProvider.Tokens(moduleCodeString);
+            _parser = new VBAParser(_tokenStream);
+            _parser.AddErrorListener(new MainParseExceptionErrorListener(moduleName, ParsePass.CodePanePass));
+            _moduleName = moduleName;
+            _mode = mode;
+        }
+
+        public (IParseTree parseTree, TokenStreamRewriter rewriter) Parse()
```

Comment:
I don't think there's a use case for parameterless call here, is there?

Reply:
That's a good point, actually. Recall the first version had it all in the constructor. To make it more usable, I guess we should at least move the `string moduleCodeString` over, so that it can be used to parse a modified module. But if we do that, why not just have it take an equivalent of `(string moduleName, string moduleCodeString)` which would means we can use a single instance of parser for more than one code module. Really, the class should be more interested in setting up the appropriate parsing strategy than whether it's handling a particular module. I'll make the change.